spec/compiler/semantic/automatic_cast_spec.cr

```diff
@@ -596,4 +635,35 @@ describe "Semantic: automatic cast" do
       ),
       "ambiguous call, implicit cast of 255 matches all of UInt64, Int64"
   end
+
+  it "casts integer variable to larger type for variable with type restriction (#9565)" do
+    assert_type(%(
+      a : Int64 = 0_i64
+      x = 1
+      a = x
```

Comment:
I didn't implement this because autocasting in this situation is a bit different. I will try to do it, but I honestly don't think it's very useful, or put another way: nobody puts type annotations in variables. But I'll try...

Reply:
Oh, sorry, I injected this spec because I initially thought about implementing it, but it depended on another PR that wasn't yet merged at that point.

I'll try to make it work again.

Reply:
Actually, let me just remove the spec.

This is very hard to implement. The reason is that when the compiler sees `x = 1` it checks whether that can be autocasted, and since "1" can't change type it's very easy to know.

With `x = a`, the type of `a` might later change... but right now the check for "can it be autocast" happens once, and it's very hard to know that... maybe `a` doesn't have a type at that point. Maybe it does and it will change later on.

Can we leave this "feature" for a separate PR, or even after 1.0? I don't think this is necessary to enjoy the real benefits of this PR, which are much more common (method calls).

Reply:
Ok. I'm fine not covering this.

Reply:
Yay! 🎉